File: backend/schema/user_stats.go
package schema

import (
	"log"

	"github.com/pocketbase/pocketbase/core"
)

// EnsureUserStatsCollection ensures the user_stats collection exists.
// 1 user = 1 record (upsert on create)
func EnsureUserStatsCollection(app core.App) {
	usersCollection, err := app.FindCollectionByNameOrId("users")
	if err != nil {
		log.Printf("Users collection not found, skipping user_stats creation")
		return
	}

	collection, err := app.FindCollectionByNameOrId("user_stats")
	if err != nil {
		collection = core.NewBaseCollection("user_stats")
	}

	changes := false
	if SetRules(collection,
		"user = @request.auth.id",
		"user = @request.auth.id",
		"user = @request.auth.id",
		"user = @request.auth.id",
		"user = @request.auth.id") {
		changes = true
	}

	if collection.Fields.GetByName("user") == nil {
		collection.Fields.Add(&core.RelationField{
			Name:          "user",
			CollectionId:  usersCollection.Id,
			Required:      true,
			MaxSelect:     1,
			CascadeDelete: true,
		})
		changes = true
	}
	if AddNumberField(collection, "total_xp", false, Ptr(0.0), nil) {
		changes = true
	}
	if AddNumberField(collection, "level", false, Ptr(1.0), Ptr(18.0)) {
		changes = true
	}
	if AddNumberField(collection, "streak_days", false, Ptr(0.0), nil) {
		changes = true
	}
	if AddTextField(collection, "last_activity_date", false) {
		changes = true
	}
	if AddNumberField(collection, "chapters_read", false, Ptr(0.0), nil) {
		changes = true
	}
	if AddNumberField(collection, "chapters_listened", false, Ptr(0.0), nil) {
		changes = true
	}
	if AddNumberField(collection, "stories_completed", false, Ptr(0.0), nil) {
		changes = true
	}

	if AddSystemFields(collection) {
		changes = true
	}

	if EnsureIndex(collection, "idx_user_stats_user", true, "user", "") {
		changes = true
	}

	if changes {
		SaveCollection(app, collection)
	}
}

File: frontend/lib/data/models/user_stats.dart
import 'package:pocketbase/pocketbase.dart';

import '../../core/constants/sticker_constants.dart';

class UserStats {
  final String id;
  final String userId;
  final double totalXp;
  final int level;
  final int streakDays;
  final String? lastActivityDate;
  final int chaptersRead;
  final int chaptersListened;
  final int storiesCompleted;

  UserStats({
    required this.id,
    required this.userId,
    this.totalXp = 0,
    this.level = 1,
    this.streakDays = 0,
    this.lastActivityDate,
    this.chaptersRead = 0,
    this.chaptersListened = 0,
    this.storiesCompleted = 0,
  });

  /// Create from local-computed values (guest, no PocketBase)
  factory UserStats.fromLocal({
    required double totalXp,
    required int chaptersRead,
    int chaptersListened = 0,
    int storiesCompleted = 0,
    int streakDays = 0,
    String? lastActivityDate,
  }) {
    final level = levelFromXp(totalXp).clamp(1, 18);
    return UserStats(
      id: 'local_guest',
      userId: 'guest',
      totalXp: totalXp,
      level: level,
      streakDays: streakDays,
      lastActivityDate: lastActivityDate,
      chaptersRead: chaptersRead,
      chaptersListened: chaptersListened,
      storiesCompleted: storiesCompleted,
    );
  }

  factory UserStats.fromRecord(RecordModel record) {
    return UserStats(
      id: record.id,
      userId: record.getStringValue('user'),
      totalXp: (record.data['total_xp'] as num?)?.toDouble() ?? 0,
      level: ((record.data['level'] as num?)?.toInt() ?? 1).clamp(1, 18),
      streakDays: (record.data['streak_days'] as num?)?.toInt() ?? 0,
      lastActivityDate: record.data['last_activity_date']?.toString(),
      chaptersRead: (record.data['chapters_read'] as num?)?.toInt() ?? 0,
      chaptersListened: (record.data['chapters_listened'] as num?)?.toInt() ?? 0,
      storiesCompleted: (record.data['stories_completed'] as num?)?.toInt() ?? 0,
    );
  }
}
